### PR TITLE
I18N-1220 Migrate to Spring boot 3.2.7

### PR DESCRIPTION
**Pull request by DarKhaos**
Migrate Mojito from Spring Boot 3.1.5 to 3.2.7

### CODE REVIEW

File: webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java [hunk 1]
```diff
@@ -150,7 +150,7 @@ public void testDeleteScreenshotNoThirdPartyConfig() throws RepositoryNameAlread
     tmTextUnitRepository.save(tmTextUnit);
 
     ScreenshotTextUnit screenshotTextUnit = new ScreenshotTextUnit();
-    screenshotTextUnit.setId(1L);
+    screenshotTextUnit.setId(2L);
```
Comment (maallen) [inline webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java:153]: Why is this change needed?

File: webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java [hunk 2]
```diff
@@ -134,7 +134,7 @@ public void testDeleteScreenshotNoThirdPartyConfig() throws RepositoryNameAlread
     screenshotService.thirdPartyService = thirdPartyServiceMock;
 
     Screenshot screenshot = new Screenshot();
-    screenshot.setId(1L);
+    screenshot.setId(2L);
```
Comment (maallen) [inline webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java:137]: Why is this change needed to the test?

File: webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java [hunk 3]
```diff
@@ -489,6 +489,7 @@ public void writeTextUnit(ITextUnit tu, String phaseName) {
             params.getPlaceholderMode(),
             params.getIncludeCodeAttrs(),
             params.getIncludeIts(),
+            false,
```
Comment (maallen) [inline webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java:492]: What is this new boolean parameter?

File: webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java [hunk 4]
```diff
@@ -505,6 +506,7 @@ public void writeTextUnit(ITextUnit tu, String phaseName) {
               params.getPlaceholderMode(),
               params.getIncludeCodeAttrs(),
               params.getIncludeIts(),
+              false,
```
Comment (maallen) [inline webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java:509]: Same as above

File: webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java [hunk 5]
```diff
@@ -581,6 +583,7 @@ public void writeTextUnit(ITextUnit tu, String phaseName) {
                 params.getPlaceholderMode(),
                 params.getIncludeCodeAttrs(),
                 params.getIncludeIts(),
+                false,
```
Comment (maallen) [inline webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java:586]: Same as above

File: webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java [hunk 6]
```diff
@@ -716,6 +719,7 @@ private void writeAltTranslations(AltTranslationsAnnotation ann, Segment segment
...
```
Comment (maallen) [inline webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java:722]: Same as above

File: webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java [hunk 7]
```diff
@@ -746,6 +750,7 @@ private void writeAltTranslations(AltTranslationsAnnotation ann, Segment segment
               params.getPlaceholderMode(),
               params.getIncludeCodeAttrs(),
               params.getIncludeIts(),
+              false,
```
Comment (maallen) [inline webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java:753]: Same as above

File: pom.xml [hunk 8]
```diff
@@ -22,7 +22,7 @@
         <maven.compiler.version>3.11.0</maven.compiler.version>
         <maven.deploy.version>2.8.2</maven.deploy.version>
         <maven.release.version>2.5.3</maven.release.version>
-        <okapi.version>1.45.0</okapi.version>
+        <okapi.version>1.46.0</okapi.version>
```
Comment (maallen) [inline pom.xml:25]: We need to test that output from push & pull commands is identical across different versions after updating okapi

File: webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java (hunk 1 above)
Comment (DarKhaos) [reply webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java:153]: This test was failing with that id

File: webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java (hunk 2 above)
Comment (DarKhaos) [reply webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java:137]: Same reason as above

File: webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java (hunk 3 above)
Comment (DarKhaos) [reply webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java:492]: it's named as: isTarget. In the new version of okapi, that parameter is used to modify the result of that method
Screenshot 2024-07-01 at 10 43 05 AM
we get the same output with isTarget = false

File: pom.xml (hunk 8 above)
Comment (DarKhaos) [reply pom.xml:25]: I have tested it against mojito dev, and it is working as expected.

File: webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java (hunk 3 above)
Comment (DarKhaos) [reply webapp/src/main/java/com/box/l10n/mojito/okapi/XLIFFWriter.java:492]: As the comment, it's for supporting XLIFF 2.0

File: webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java (hunk 2 above)
Comment (DarKhaos) [reply webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java:137]: Before the changes, the id of the screenshot variable was updated to 29. After the changes, the id keeps as 1 (the record with that id was removed here:).
Screenshot 2024-07-01 at 9 08 49 PM

File: webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java (hunk 1 above)
Comment (DarKhaos) [reply webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java:153]: Before the changes, the screenshot variable had an ID of 29. However, after the changes, its ID remains 1 (the record with that id was removed here and it's causing the issue).
Screenshot 2024-07-01 at 9 08 49 PM

Comment (maallen) [reply webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java:153]: Not sure I follow? The test deletes a screenshot and verifies it has been deleted, what error do you see when it's left as is?

File: webapp/src/main/java/com/box/l10n/mojito/service/blobstorage/database/MBlobRepository.java [hunk 9]
```diff
@@ -31,7 +31,7 @@ public interface MBlobRepository
   @Query(
       """
       select mb.id from #{#entityName} mb
-      where (unix_timestamp(mb.createdDate) + mb.expireAfterSeconds) < unix_timestamp(:now)
+      where (cast(unix_timestamp(mb.createdDate) as biginteger) + cast(mb.expireAfterSeconds as biginteger)) < cast(unix_timestamp(:now) as biginteger)
```
Comment (maallen) [inline webapp/src/main/java/com/box/l10n/mojito/service/blobstorage/database/MBlobRepository.java:34]: What error were you seeing that required the cast?

Comment (DarKhaos) [reply webapp/src/main/java/com/box/l10n/mojito/service/blobstorage/database/MBlobRepository.java:34]: This is the error:

> Caused by: java.lang.IllegalArgumentException: org.hibernate.query.SemanticException: Operand of + is of type 'java.lang.Object' which is not a numeric type (it is not an instance of 'java.lang.Number', 'java.time.Temporal', or 'java.time.TemporalAmount')
> 	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:143)
> 	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:167)
> 	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:173)
> 	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:848)
> 	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:753)
> 	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:136)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
> 	at org.springframework.orm.jpa.ExtendedEntityManagerCreator$ExtendedEntityManagerInvocationHandler.invoke(ExtendedEntityManagerCreator.java:364)
> 	at jdk.proxy2/jdk.proxy2.$Proxy218.createQuery(Unknown Source)
> 	at org.springframework.data.jpa.repository.query.SimpleJpaQuery.validateQuery(SimpleJpaQuery.java:94)
> 	... 110 common frames omitted
> Caused by: org.hibernate.query.SemanticException: Operand of + is of type 'java.lang.Object' which is not a numeric type (it is not an instance of 'java.lang.Number', 'java.time.Temporal', or 'java.time.TemporalAmount')
> 	at org.hibernate.query.sqm.internal.TypecheckUtil.assertOperable(TypecheckUtil.java:490)
> 	at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitAdditionExpression(SemanticQueryBuilder.java:2958)
> 	at org.hibernate.grammars.hql.HqlParser$AdditionExpressionContext.accept(HqlParser.java:7114)
> 	at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitGroupedExpression(SemanticQueryBuilder.java:3038)
> 	at org.hibernate.grammars.hql.HqlParser$GroupedExpressionContext.accept(HqlParser.java:7232)
> 	at org.hibernate.query.hql.internal.SemanticQueryBuilder.createComparisonPredicate(SemanticQueryBuilder.java:2429)
> 	at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitComparisonPredicate(SemanticQueryBuilder.java:2392)
> 	at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitComparisonPredicate(SemanticQueryBuilder.java:269)
> 	at org.hibernate.grammars.hql.HqlParser$ComparisonPredicateContext.accept(HqlParser.java:6165)
> 	at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitWhereClause(SemanticQueryBuilder.java:2244)
> 	at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitWhereClause(SemanticQueryBuilder.java:269)
> 	at org.hibernate.grammars.hql.HqlParser$WhereClauseContext.accept(HqlParser.java:5906)
> 	at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitQuery(SemanticQueryBuilder.java:1159)
> 	at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitQuerySpecExpression(SemanticQueryBuilder.java:941)
> 	at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitQuerySpecExpression(SemanticQueryBuilder.java:269)
> 	at org.hibernate.grammars.hql.HqlParser$QuerySpecExpressionContext.accept(HqlParser.java:1869)
> 	at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitSimpleQueryGroup(SemanticQueryBuilder.java:926)
> 	at org.hibernate.query.hql.internal.SemanticQueryBuilder.visitSimpleQueryGroup(SemanticQueryBuilder.java:269)

File: webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java (hunk 1 above)
Comment (DarKhaos) [reply webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java]: For the current changes, the code below prints "1 - 29" for test **testDeleteScreenshotNoThirdPartyConfig**
Screenshot 2024-07-02 at 11 44 50 AM
Before the changes, it printed "29 - 29". That record (with ID = 1) was removed in test testDeleteScreenshot, and that is why the assert `Assert.assertNotNull(screenshotRepository.findById(screenshot.getId()).orElse(null));` threw an exception

Comment (DarKhaos) [reply webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java]: This was changed to the following:
Screenshot 2024-07-03 at 7 28 15 PM

Comment (maallen) [reply webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java]: Is your database being cleaned before the test execution? It sounds like the id mismatch is due to you already having screenshots existing in the test database before the test is run. Below  in your application.properties file will clean the test database prior to test execution.

Ensure to comment it out when running some manual testing locally as it will clean your dev db in that scenario


`l10n.flyway.clean=true
spring.flyway.cleanDisabled=false
spring.flyway.clean-disabled=false`

Comment (maallen) [reply webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java]: My issue here is finding out if this is caused by the Spring upgrade or is it purely environmental, if it's the Spring upgrade what's changed to cause this behaviour and will it affect us elsewhere?

Comment (maallen) [reply webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java]: @DarKhaos missed this was outdated, new code looks fine (running the tests against a fresh db locally is important though)

**Review by maallen — APPROVED** [review]

**Comment by maallen** [comment]
Code looks fine, have you verified that the generated localized assets are the same before/after the migration?

File: webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java [hunk 10]
```diff
@@ -134,9 +134,8 @@ public void testDeleteScreenshotNoThirdPartyConfig() throws RepositoryNameAlread
     screenshotService.thirdPartyService = thirdPartyServiceMock;
 
     Screenshot screenshot = new Screenshot();
-    screenshot.setId(1L);
```
Comment (DarKhaos) [inline webapp/src/test/java/com/box/l10n/mojito/service/screenshot/ScreenshotServiceTest.java:137]: See Hibernate bug: https://hibernate.atlassian.net/browse/HHH-17634
And this is the PR: https://github.com/hibernate/hibernate-orm/pull/7698

File: pom.xml (hunk 8 above)
Comment (DarKhaos) [inline pom.xml:25]: TheOkapi framework was updated because it doesn't support SnakeYAML 2.2 (which is updated by Spring Boot 3.2.7) 
https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.2-Release-Notes

**Comment by DarKhaos** [comment]
> Code looks fine, have you verified that the generated localized assets are the same before/after the migration?

Yes, I have verified the localized files
